%!TEX root = ../GEtest-draft.tex

\section{Introduction}\label{sec:intro}

In traditional genome-wide association studies (GWAS), the detection of genetic association between a single-nucleotide polymorphism (SNP, G) and a quantitative or binary trait (Y) is often done by testing on the main effect of G alone. This framework will have inferior power if the true casual SNPs have trivial main effects but strong interaction effects with some environmental variable (E). A known example of such SNP is rs12753193(LEPR), which has been shown to have a strong interaction effect with BMI on C-reactive protein levels but no detectable main effect \citep{lepr}.

The main reason that accounting gene-environment interaction effect (GxE) is generally difficult in most GWAS studies is information from the interacting environmental variable E is often missing or collected with non-trivial measurement error \citep{jlst}. For quantitative traits (Y) that are approximately normally distributed, it has been shown that ignoring the environmental variable E when it is interacting with a SNP will produce an artificial heteroskedasticity of Y across genotypes \citep{lepr}. For this reason, indirect testing methods that detect the GxE interaction by testing the heteroskedasticity of Y have been proposed for quantitative traits \citep{jlst,gjlst}. The indirect testing method enables the detection of significant GxE interaction even if there is no information collected on variable E, and hence allows its effect to be accounted for easily for studies of genetic association between a SNP with a quantitative trait.

However, the approach of testing potential GxE interaction by checking variance of traits across groups will not generalize to the case where traits of interest are binary such as disease affection status, because the variance of binary variable is directly specified by its mean. This places a constraint on how to account for the GxE interaction effect in a GWAS study for a binary trait.

In this paper, we propose a novel methodology for indirect testing of GxE interaction for binary traits by exploring the latent variable framework of the probit regression model. Using the latent variable framework, we will first show that for a SNP with additive main effect, ignoring the GxE interaction will result in a non-additive genotypic model, and hence propose a method to detect the GxE interaction without using the information of E. Then we will show that under the existence of a suitable auxiliary variable Z, the power of our proposed method can be improved, and the proposed method can be generalized to the case where the main effect of SNP is genotypic instead of additive.

The remainder of this paper is organized as follows. In section \ref{sec:prelim}, we will give more details on the latent variable formulation of logistic/probit regression, and discuss the details of why methods based on the Levene test will not work in this setting. In section \ref{sec:methods}, we will describe our novel methodology for indirect testing of GxE interaction, and explain how it can be improved if there exists suitable auxiliary variable Z. In section \ref{sec:simulations}, we will show that our proposed approach has a well-controlled type I error rate and satisfactory power even compared to the direct testing method with information of E available, through an extensive set of simulation studies using the 1000 Genome Project (1kGP) dataset \citep{1KGP}. In section \ref{sec:examples}, we will implement our proposed method on the UK Biobank (UKB) dataset, to identify promising SNPs that are likely to be interacting with some unknown environmental variables \citep{UKB}. We conclude with a discussion in section \ref{sec:discussion}.



