% Options for packages loaded elsewhere
\PassOptionsToPackage{unicode}{hyperref}
\PassOptionsToPackage{hyphens}{url}
%
\documentclass[
]{article}
\usepackage{lmodern}
\usepackage{amssymb,amsmath}
\usepackage{ifxetex,ifluatex}
\ifnum 0\ifxetex 1\fi\ifluatex 1\fi=0 % if pdftex
  \usepackage[T1]{fontenc}
  \usepackage[utf8]{inputenc}
  \usepackage{textcomp} % provide euro and other symbols
\else % if luatex or xetex
  \usepackage{unicode-math}
  \defaultfontfeatures{Scale=MatchLowercase}
  \defaultfontfeatures[\rmfamily]{Ligatures=TeX,Scale=1}
\fi
% Use upquote if available, for straight quotes in verbatim environments
\IfFileExists{upquote.sty}{\usepackage{upquote}}{}
\IfFileExists{microtype.sty}{% use microtype if available
  \usepackage[]{microtype}
  \UseMicrotypeSet[protrusion]{basicmath} % disable protrusion for tt fonts
}{}
\makeatletter
\@ifundefined{KOMAClassName}{% if non-KOMA class
  \IfFileExists{parskip.sty}{%
    \usepackage{parskip}
  }{% else
    \setlength{\parindent}{0pt}
    \setlength{\parskip}{6pt plus 2pt minus 1pt}}
}{% if KOMA class
  \KOMAoptions{parskip=half}}
\makeatother
\usepackage{xcolor}
\IfFileExists{xurl.sty}{\usepackage{xurl}}{} % add URL line breaks if available
\IfFileExists{bookmark.sty}{\usepackage{bookmark}}{\usepackage{hyperref}}
\hypersetup{
  pdftitle={Progress report: Detecting interaction with unknown environmental covariate},
  pdfauthor={Ziang Zhang},
  hidelinks,
  pdfcreator={LaTeX via pandoc}}
\urlstyle{same} % disable monospaced font for URLs
\usepackage[margin=1in]{geometry}
\usepackage{color}
\usepackage{fancyvrb}
\newcommand{\VerbBar}{|}
\newcommand{\VERB}{\Verb[commandchars=\\\{\}]}
\DefineVerbatimEnvironment{Highlighting}{Verbatim}{commandchars=\\\{\}}
% Add ',fontsize=\small' for more characters per line
\usepackage{framed}
\definecolor{shadecolor}{RGB}{248,248,248}
\newenvironment{Shaded}{\begin{snugshade}}{\end{snugshade}}
\newcommand{\AlertTok}[1]{\textcolor[rgb]{0.94,0.16,0.16}{#1}}
\newcommand{\AnnotationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\AttributeTok}[1]{\textcolor[rgb]{0.77,0.63,0.00}{#1}}
\newcommand{\BaseNTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\BuiltInTok}[1]{#1}
\newcommand{\CharTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\CommentTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\CommentVarTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ConstantTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ControlFlowTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\DataTypeTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{#1}}
\newcommand{\DecValTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\DocumentationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\ErrorTok}[1]{\textcolor[rgb]{0.64,0.00,0.00}{\textbf{#1}}}
\newcommand{\ExtensionTok}[1]{#1}
\newcommand{\FloatTok}[1]{\textcolor[rgb]{0.00,0.00,0.81}{#1}}
\newcommand{\FunctionTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\ImportTok}[1]{#1}
\newcommand{\InformationTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\newcommand{\KeywordTok}[1]{\textcolor[rgb]{0.13,0.29,0.53}{\textbf{#1}}}
\newcommand{\NormalTok}[1]{#1}
\newcommand{\OperatorTok}[1]{\textcolor[rgb]{0.81,0.36,0.00}{\textbf{#1}}}
\newcommand{\OtherTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{#1}}
\newcommand{\PreprocessorTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textit{#1}}}
\newcommand{\RegionMarkerTok}[1]{#1}
\newcommand{\SpecialCharTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\SpecialStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\StringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\VariableTok}[1]{\textcolor[rgb]{0.00,0.00,0.00}{#1}}
\newcommand{\VerbatimStringTok}[1]{\textcolor[rgb]{0.31,0.60,0.02}{#1}}
\newcommand{\WarningTok}[1]{\textcolor[rgb]{0.56,0.35,0.01}{\textbf{\textit{#1}}}}
\usepackage{graphicx,grffile}
\makeatletter
\def\maxwidth{\ifdim\Gin@nat@width>\linewidth\linewidth\else\Gin@nat@width\fi}
\def\maxheight{\ifdim\Gin@nat@height>\textheight\textheight\else\Gin@nat@height\fi}
\makeatother
% Scale images if necessary, so that they will not overflow the page
% margins by default, and it is still possible to overwrite the defaults
% using explicit options in \includegraphics[width, height, ...]{}
\setkeys{Gin}{width=\maxwidth,height=\maxheight,keepaspectratio}
% Set default figure placement to htbp
\makeatletter
\def\fps@figure{htbp}
\makeatother
\setlength{\emergencystretch}{3em} % prevent overfull lines
\providecommand{\tightlist}{%
  \setlength{\itemsep}{0pt}\setlength{\parskip}{0pt}}
\setcounter{secnumdepth}{5}

\title{Progress report: Detecting interaction with unknown environmental
covariate}
\author{Ziang Zhang}
\date{15/10/2020}

\begin{document}
\maketitle

\hypertarget{the-underlying-model}{%
\section{The Underlying Model:}\label{the-underlying-model}}

For binary response variable, it is often assumed that the response
variable \(y_i\) conditioning on the regressors \(G_i,Z_i\) come from a
latent model such that: \begin{equation}\label{eqn:latentformulation}
\begin{aligned}
Y_i^* &= \beta_0 + \beta_G G_i + \beta_Z Z_i + \epsilon_i \\
Y_i &= I\{Y_i^*>0\} \\
\end{aligned}
\end{equation}

The unobserved latent variable \(Y_i^*\) determines whether the observed
response variable \(Y_i\) is 0 or 1. The error term \(\epsilon_i\) in
\(Y_i^*\) needs to have a completely known distribution, which can be
\(\text{N}(0,1)\) for the model of \(Y|G,Z\) to become a probit model,
or a logistic distribution with mean 0 and variance 3.28 for the model
to become a logistic regression model.

Here the regressor \(G_i\) represents the allele of interest, and the
regressor \(Z_i\) is any regressor that can be non-genetic. For now on,
we will assume the model is probit for simplicity, unless otherwise
indicated.

Similarly, we can have a Genotypic Model defined as:\\
\begin{equation}\label{eqn:latentformulationGeno}
\begin{aligned}
Y_i^* &= \beta_0 + \beta_{G1} I(G_i = 1) + \beta_{G2} I(G_i = 2) + \beta_Z Z_i + \epsilon_i \\
Y_i &= I\{Y_i^*>0\} \\
\end{aligned}
\end{equation}

The Genotypic Model has higher degree of freedom than the additive model
due to the extra regression parameter.

\hypertarget{when-the-true-model-does-contain-gene-environment-interaction}{%
\subsection{When the true model does contain gene-environment
interaction}\label{when-the-true-model-does-contain-gene-environment-interaction}}

Assume for simplicity that \(E_i\) the environmental variable has a
normal distribution with mean \(\mu_E\) and variance \(\sigma_E^2\), and
suppose that the true underlying model is:
\begin{equation}\label{eqn:probitModelWithInteraction}
\begin{aligned}
Y_i^* &= \beta_0 + \beta_G G_i + \beta_Z Z_i + \beta_E E_i + \beta_{G\times E} G_i \times E_i + \epsilon_i \\
Y_i &= I\{Y_i^*>0\} \\
\epsilon_i &\sim \text{N}(0,1)
\end{aligned}
\end{equation}

Furthermore, we can compute that:
\begin{equation}\label{eqn:probitModelWithInteraction_MeanVar}
\begin{aligned}
\text{E}(Y_i^*|G_i,Z_i) &= \beta_0 + \beta_E \mu_E + (\beta_G + \beta_{G\times E} \mu_E)G_i + \beta_Z Z_i \\
\text{Var}(Y_i^*|G_i,Z_i) &= (\beta_{G\times E} G_i)^2 \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1 \\
Y_i^*|G_i, Z_i &\sim \text{N}\bigg(\beta_0 + \beta_E \mu_E + (\beta_G + \beta_{G\times E} \mu_E)G_i + \beta_Z Z_i, (\beta_{G\times E} G_i)^2 \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1 \bigg)
\end{aligned}
\end{equation}

That implies that the probability we get a case for different levels of
\(G_i\) and \(Z_i\) will be:
\begin{equation}\label{eqn:probitModelWithInteraction_Prob} 
\begin{aligned} 
\text{P}(Y = 1 | G_i, Z_i) &= \text{P}(Y^* > 0| G_i, Z_i) \\ 
                           &= \text{P}(\frac{Y^*  - \text{E}(Y^* |G_i,Z_i)}{\sqrt{\text{Var}(Y^* |G_i,Z_i)}} > \frac{-\text{E}(Y^* |G_i,Z_i)}{\sqrt{\text{Var}(Y^* |G_i,Z_i)}}) \\
                           &= \Phi \bigg( \frac{\text{E}(Y^* |G_i,Z_i)}{\sqrt{\text{Var}(Y^* |G_i,Z_i)}} \bigg)
\end{aligned}
\end{equation}

Therefore, applying the inverse CDF on both sides, we get
\[\Phi^{-1} \bigg(\text{P}(Y = 1 | G, Z) \bigg) = \frac{\beta_0+\beta_E \mu_E+(\beta_G + \beta_{G\times E} \mu_E)G_i + \beta_Z Z}{\sqrt{(\beta_{G\times E}^2 G_i^2 \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1)}} \]

This is only a linear function of \(G_i\) when the interaction parameter
\(\beta_{G\times E} = 0\), and the slope of Z is constant across
different genes only when \(\beta_{G\times E}\) is zero.

\begin{enumerate}
\item If the true underlying model also contains another regressor $W$ but $W$ is uncorrelated with $G$ for example. Then even though ignoring that regressor breaks the structural assumption of probit model, so that the fitted model without $W$ is no longer a probit model (since now $\epsilon$ does not follow standard normal), but $\Phi^{-1}(\text{P}(Y_i = 1|G_i,Z_i))$ will still be a linear function of $G_i$. So detecting based on the linearity of $\Phi^{-1}\text{P}$ will not be affected by omitted exogenous regressors.
\item Since $P(Y_i = 1|G_i,Z_i)$ is actually unknown in practice, we can estimate it using the sample proportion $\hat{P}(Y = 1|G = g,Z = z) = \frac{\sum_{i=1}^{n} \text{I}\{y_i =1,G_{i} = g, Z_{i} = z\}}{\sum_{i=1}^{n}  \text{I}\{G_{i} = g, Z_{i} = z\}}$. We shouldn't use the fitted model to estimate them since our fitted model may be wrong.
\item The reason we used probit model instead of logistic model here is that assuming $E$ follows normal distribution, $Y^*|G,Z$ will still be normal if we omit the interaction term, since linear combination of normal is normal. But assuming $E$ follows logistic distribution does not imply that $Y^*|G,Z$ will be logistically distributed as logistic distribution is not closed under linear combination. However, logistic regression model for $Y|G,Z$ implies that $Y^*|G,Z$ must follow logistic distribution. In other words, probit regression model with omitted covariate will still be a probit regression model, just with different regression parameters. Based on the literature, it seems like probit model and logistic model have really closed results in real applications.
\end{enumerate}

If the model is Genotypic instead: \begin{equation}\label{eqn:genointer}
\begin{aligned}
Y_i^* &= \beta_0 + \beta_{G1} I(G_i = 1) + \beta_{G2} I(G_i = 2) + \beta_Z Z_i + \beta_E E_i + \beta_{G1E} I(G_i = 1) \times E_i + \beta_{G2E} I(G_i = 2) \times E_i  + \epsilon_i \\
\end{aligned}
\end{equation}

then we can derive the following:
\[\Phi^{-1} \bigg(\text{P}(Y = 1 | G, Z) \bigg) = \frac{\beta_0+\beta_E \mu_E+(\beta_{G1} + \beta_{G1E} \mu_E)I(G = 1)+(\beta_{G2} + \beta_{G2E} \mu_E)I(G = 2) + \beta_Z Z}{\sqrt{(\beta_{G1E}^2 I(G = 1) \sigma_E^2 +\beta_{G2E}^2 I(G = 2) \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1)}} \]
In this case, the model will still be linear in \(I(G = 1), I(G = 2)\)
because of the extra parameter, just with different regression
parameters. But the slope of Z will continue to differ between different
Genetic types unless the there are no interaction effects
\(\beta_{G1E}\) and \(\beta_{G2E}\).

\clearpage

\hypertarget{method-for-additive-model}{%
\section{Method for Additive Model:}\label{method-for-additive-model}}

In this section, I will present two methods for the detection of
interaction effect when the true model is additive.

\hypertarget{testing-of-linearity}{%
\subsection{Testing of Linearity:}\label{testing-of-linearity}}

Recall that when the model is additive, then:
\[\Phi^{-1} \bigg(\text{P}(Y = 1 | G, Z) \bigg) = \frac{\beta_0+\beta_E \mu_E+(\beta_G + \beta_{G\times E} \mu_E)G + \beta_Z Z}{\sqrt{(\beta_{G\times E}^2 G^2 \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1)}}\]

This method relies on the checking of linearity of \(\Phi^{-1}(P)\), so
the test statistics will also be focusing on the detection of linearity.
Note that the above 1 degree of freedom regression model, although it is
not linear in \(G\), it can be rewritten as a valid 2 degree of freedom
genotypic model that is linear in \(I(G=1), I(G=2)\):

\begin{equation}
\begin{aligned}
\Phi^{-1} \bigg(\text{P}(Y = 1 | G, Z) \bigg) &= \frac{\beta_0+\beta_E \mu_E+(\beta_G + \beta_{G\times E} \mu_E)G + \beta_Z Z}{\sqrt{(\beta_{G\times E}^2 G^2 \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1)}} \\
                                              &= \gamma_0 + \gamma_1 I(G = 1) + \gamma_2 I(G=2) + \gamma_{Z1G} I(G=1) * Z + \gamma_{Z2G} I(G=2) * Z + \gamma_Z Z 
\end{aligned}
\end{equation}

Here, we can compute that: \begin{equation}
\begin{aligned}
& \gamma_0 = \frac{\beta_0 + \beta_E \mu_E}{\sqrt{\beta_E^2 \sigma_E^2 + 1}} \\
& \gamma_1 = \frac{\beta_0 + \beta_E \mu_E + \beta_G + \beta_{G\times E} \mu_E}{\sqrt{\beta_{G\times E}^2 \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1}} - \gamma_0 \\
& \gamma_2 = \frac{\beta_0 + \beta_E \mu_E + 2(\beta_G + \beta_{G\times E} \mu_E)}{\sqrt{4\beta_{G\times E}^2 \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1}} - \gamma_0
\end{aligned}
\end{equation}

Therefore, we can conclude that if \(\beta_{G\times E} = 0\), then
\(\beta_G =\gamma_1 = 2\gamma_2\) must hold. If we further have the
information on the covariate \(Z\), then we can increase the power of
our detection by also testing on \(\gamma_{Z1G} = \gamma_{Z2G} =0\)
(equal slopes of Z across genotypes).

\hypertarget{wald-test-statistics}{%
\subsubsection{Wald Test Statistics:}\label{wald-test-statistics}}

To test the null hypothesis of \(\beta_{G\times E} = 0\), we can use the
following steps:

\textbf{1. Rewrite the additive regression model:} We first rewrite the
additive regression model as a genotypic model, i.e:
\[\Phi^{-1} \bigg(\text{P}(Y = 1 | G, Z) \bigg) = \gamma_0 + \gamma_1 I(G = 1) + \gamma_2 I(G=2) + \gamma_{Z1G} I(G=1) * Z + \gamma_{Z2G} I(G=2) * Z + \gamma_Z Z\]

Now, the regression parameters in this model may not carry very
meaningful interpretations due to the potential presence of
\(\beta_{G\times E}\). However, we can use them to detect the presence
of missing interaction effect by either testing
\(H_0: \gamma_1 = \gamma_2\) or
\(H_0: \gamma_1 = \gamma_2, \ \gamma_{Z1G}=\gamma_{Z2G} =0\), depending
on whether the information of \(Z\) is available, and whether \(Z\) is
ordinal.

\textbf{2. Wald Test on linearity:} To test \(H_0: \gamma_1 = \gamma_2\)
or \(H_0: \gamma_1 = \gamma_2, \ \gamma_{Z1G}=\gamma_{Z2G} =0\), we can
simply do a Wald test on the genotypic working model that we considered
above. It turns out that this Wald test statistic \(T\) can also be
equivalently derived through an application of Delta method on sample
proportion, or as a two-stage linear regression.

If we have information about the covariate \(Z\) in the regression
model, and \(Z\) is an ordinal variable with additive effect, then the
power of our test will be augmented if we test
\(H_0: \gamma_1 = \gamma_2, \ \gamma_{Z1G}=\gamma_{Z2G} =0\) in our Wald
test. We can also only test \(H_0: \gamma_{Z1G}=\gamma_{Z2G} =0\), this
makes our test robust to the case that G's effect is actually
non-additive. However, we then need to be more careful to make sure that
\(Z\) indeed has no interaction with G or E in the \textbf{true} model.

\hypertarget{method-for-genotypic-model}{%
\section{Method for Genotypic Model:}\label{method-for-genotypic-model}}

\hypertarget{auxiliary-variable-method}{%
\subsection{Auxiliary variable
method:}\label{auxiliary-variable-method}}

Recall for a Genotypic Model with interaction like below:
\begin{equation}\label{eqn:genointer}
\begin{aligned}
Y_i^* &= \beta_0 + \beta_{G1} I(G_i = 1) + \beta_{G2} I(G_i = 2) + \beta_Z Z_i + \beta_E E_i + \beta_{G1E} I(G_i = 1) \times E_i + \beta_{G2E} I(G_i = 2) \times E_i  + \epsilon_i \\
\end{aligned}
\end{equation}

We can derive that: \begin{equation}\label{eqn:RStest}
\begin{aligned}
\Phi^{-1} \bigg(\text{P}(Y = 1 | G, Z) \bigg) &= \frac{\beta_0+\beta_E \mu_E+(\beta_{G1} + \beta_{G1E} \mu_E)I(G = 1)+(\beta_{G2} + \beta_{G2E} \mu_E)I(G = 2) + \beta_Z Z}{\sqrt{(\beta_{G1E}^2 I(G = 1) \sigma_E^2 +\beta_{G2E}^2 I(G = 2) \sigma_E^2 + \beta_E^2 \sigma_E^2 + 1)}} \\
&= \gamma_0 + \gamma_1 I(G = 1) + \gamma_2 I(G = 2) + \gamma_Z Z + \gamma_{Z1G} I(G = 1) \times Z + \gamma_{Z2G} I(G = 2) \times Z
\end{aligned}
\end{equation}

where the new parameters \(\gamma_{Z1G}\) and \(\gamma_{Z2G}\) will be
defined as:
\[\gamma_{Z1G} = \frac{\beta_Z}{\sqrt{\beta_E^2\sigma_E^2+\beta_{G1E}^2\sigma_E^2 +1}}-\frac{\beta_Z}{\sqrt{\beta_E^2\sigma_E^2 +1}} \]
and:
\[\gamma_{Z2G} = \frac{\beta_Z}{\sqrt{\beta_E^2\sigma_E^2+\beta_{G2E}^2\sigma_E^2 +1}}-\frac{\beta_Z}{\sqrt{\beta_E^2\sigma_E^2 +1}} \]

In other words, a Genotypic Model with an missing interaction can still
be written as a linear function of these two indicator functions of G
because of the extra regression parameter. However, ignoring this
environment to gene interaction will create an artificial interaction
between gene and the covariate Z. Since the interaction effects are zero
if and only if the covariate Z has constant slopes across different
genotypes, we can test the environmental interaction by testing the null
hypothesis \(H_0: \gamma_{Z1G} = \gamma_{Z2G} = 0\), using either wald
test, likelihood ratio test or score test.

The key in this method is to test the equal slopes of the auxiliary
variable \(Z\). In order for this method to work, we need the following
assumption:

\begin{enumerate}
\item The auxiliary variable $Z_i$ is assumed to have no interaction effect with G in the model conditional on G, Z and E.
\item The auxiliary variable $Z_i$ is also assumed to have no interaction effect with E in the model conditional on G, Z and E.
\end{enumerate}

\textbf{Main concern about genotypic model:} When we use the above
method to test for the presence of GE interaction in a genotypic model,
the power of our test is highly dependent on the true values for
\(\beta_Z\) and \(\beta_{G\times E}\). Since \(\gamma_{Z1G}\) and
\(\gamma_{Z2G}\) will eventually converge to \(0\) as \(\beta_Z\) gets
closer to \(0\), or as \(\beta_{G\times E}\) gets closer to \(0\). In
other words, our method has very low power when:
\[\frac{\beta_Z}{\sqrt{\beta_E^2\sigma_E^2+\beta_{G1E}^2\sigma_E^2 +1}} \approx \frac{\beta_Z}{\sqrt{\beta_E^2\sigma_E^2+\beta_{G2E}^2\sigma_E^2 +1}} \approx \frac{\beta_Z}{\sqrt{\beta_E^2\sigma_E^2+1}}\]

\hypertarget{gwas-implementation-on-the-1000-genome-project-data}{%
\section{GWAS Implementation on the 1000 Genome Project
Data:}\label{gwas-implementation-on-the-1000-genome-project-data}}

\textbf{Classical Testing for Main effect}

In this section, we will conduct a GWAS on the 1kGP dataset. The cleaned
set of data has 1736 independent individuals and around 2 millions SNPs.
However, since our analysis will be more sensitive to the low genotypic
frequency of each gene, we will further have some additional quality
control to make sure the genotypic frequency is high enough for each
category. We further filtered SNPs in this dataset with threshold on MAF
being \(0.15\), and keep only the SNPs on autosomes. After this
additional QC procedure, there are 1749 individuals with 150652 SNPs in
our dataset.

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{### Read in data:}
\NormalTok{path <-}\StringTok{ "D:/gwas-practice/indep_QC.bed"}
\NormalTok{tmpfile  <-}\StringTok{ }\KeywordTok{tempfile}\NormalTok{()}
\KeywordTok{snp_readBed}\NormalTok{(path, }\DataTypeTok{backingfile =}\NormalTok{ tmpfile)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## [1] "C:\\Users\\aguer\\AppData\\Local\\Temp\\RtmpyWxJew\\file3c7c78c34ca2.rds"
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\NormalTok{obj.bigSNP <-}\StringTok{ }\KeywordTok{snp_attach}\NormalTok{(}\KeywordTok{paste0}\NormalTok{(tmpfile , }\StringTok{".rds"}\NormalTok{))}

\NormalTok{G   <-}\StringTok{ }\NormalTok{obj.bigSNP}\OperatorTok{$}\NormalTok{genotypes}
\NormalTok{CHR <-}\StringTok{ }\NormalTok{obj.bigSNP}\OperatorTok{$}\NormalTok{map}\OperatorTok{$}\NormalTok{chromosome}
\NormalTok{POS <-}\StringTok{ }\NormalTok{obj.bigSNP}\OperatorTok{$}\NormalTok{map}\OperatorTok{$}\NormalTok{physical.pos}

\CommentTok{# Check some counts for the 10 first SNPs}
\KeywordTok{big_counts}\NormalTok{(G, }\DataTypeTok{ind.col =} \DecValTok{1}\OperatorTok{:}\DecValTok{10}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## 0     635 1046 1273  637 1247 1249  995 1127  535  1228
## 1     774  572  423  663  465  454  590  509  796   445
## 2     340  131   53  449   37   46  164  113  418    76
## <NA>    0    0    0    0    0    0    0    0    0     0
\end{verbatim}

We randomly assigned an individual to case or control, and assume the
disease prevalence is 0.3. Firstly, we conduct the classical logistic
regression to test the main effects. Because of the random assignment,
the p-values should have similar behavior as \(\text{Unif}[0,1]\). The
results are summarized into histogram, QQ plot and Manhattan plot at
below:

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{### Randomly generate case/control data under the null hypothesis}
\KeywordTok{set.seed}\NormalTok{(}\DecValTok{123}\NormalTok{)}
\NormalTok{case <-}\StringTok{ }\KeywordTok{rbinom}\NormalTok{(}\KeywordTok{nrow}\NormalTok{(G),}\DataTypeTok{size =} \DecValTok{1}\NormalTok{,}\DataTypeTok{prob =} \FloatTok{0.3}\NormalTok{)}
\NormalTok{obj.bigSNP}\OperatorTok{$}\NormalTok{fam}\OperatorTok{$}\NormalTok{case <-}\StringTok{ }\NormalTok{case}


\CommentTok{### Testing for main effects using Logistic regression:}
\NormalTok{obj.gwas <-}\StringTok{ }\KeywordTok{big_univLogReg}\NormalTok{(G, }\DataTypeTok{y01.train =}\NormalTok{ case,}
                           \DataTypeTok{ncores =}\NormalTok{ 6L, }\DataTypeTok{maxiter =} \DecValTok{100}\NormalTok{)}


\CommentTok{### QQ plot, Manhattan plot and Genomic Control}
\KeywordTok{snp_qq}\NormalTok{(}\DataTypeTok{gwas =}\NormalTok{ obj.gwas)}
\end{Highlighting}
\end{Shaded}

\includegraphics{stats-gene-research-progress-v6_files/figure-latex/unnamed-chunk-2-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{snp_manhattan}\NormalTok{(}\DataTypeTok{gwas =}\NormalTok{ obj.gwas, }\DataTypeTok{infos.chr =}\NormalTok{ CHR, }\DataTypeTok{infos.pos =}\NormalTok{ POS)}
\end{Highlighting}
\end{Shaded}

\includegraphics{stats-gene-research-progress-v6_files/figure-latex/unnamed-chunk-2-2.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{### Histogram of p-values:}
\NormalTok{p_vals <-}\StringTok{ }\DecValTok{2}\OperatorTok{*}\KeywordTok{pnorm}\NormalTok{(}\OperatorTok{-}\KeywordTok{abs}\NormalTok{(obj.gwas}\OperatorTok{$}\NormalTok{score))}
\KeywordTok{hist}\NormalTok{(p_vals, }\DataTypeTok{breaks =} \DecValTok{30}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{stats-gene-research-progress-v6_files/figure-latex/unnamed-chunk-2-3.pdf}

Based on the plots above, we can conclude that p-value's distribution is
very close to the uniform distribution, which is what we expect.

\clearpage

\textbf{Proposed Method: Assuming additive effect}

Then, we will apply our proposed methodology for detection of missing
interaction, assuming the true effect of gene is additive. Again, we
expect to see the distribution of p-values to be close to uniform.

\begin{Shaded}
\begin{Highlighting}[]
\KeywordTok{load}\NormalTok{(}\StringTok{"D:/gwas-practice/additive_testing.Rdata"}\NormalTok{)}

\CommentTok{## View of the result}
\KeywordTok{head}\NormalTok{(result_P1)}
\end{Highlighting}
\end{Shaded}

\begin{verbatim}
## # A tibble: 6 x 5
##   SNP            CHR      BP    stats      P
##   <chr>        <int>   <int>    <dbl>  <dbl>
## 1 SNP1-840643      1  850780 0.000836 0.977 
## 2 SNP1-842827      1  852964 0.0175   0.895 
## 3 SNP1-908480      1  918617 2.39     0.122 
## 4 rs3766192        1 1017197 0.421    0.516 
## 5 rs3766191        1 1017587 4.33     0.0375
## 6 SNP1-1013008     1 1023145 1.41     0.236
\end{verbatim}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## QQ plot of p-values}
\KeywordTok{qq}\NormalTok{(}\KeywordTok{na.omit}\NormalTok{(result_P1}\OperatorTok{$}\NormalTok{P))}
\end{Highlighting}
\end{Shaded}

\includegraphics{stats-gene-research-progress-v6_files/figure-latex/unnamed-chunk-3-1.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## Manhattan plot of p-values:}
\KeywordTok{manhattan}\NormalTok{(}\KeywordTok{na.omit}\NormalTok{(result_P1))}
\end{Highlighting}
\end{Shaded}

\includegraphics{stats-gene-research-progress-v6_files/figure-latex/unnamed-chunk-3-2.pdf}

\begin{Shaded}
\begin{Highlighting}[]
\CommentTok{## Histogram of p-values:}
\KeywordTok{hist}\NormalTok{(result_P1}\OperatorTok{$}\NormalTok{P, }\DataTypeTok{breaks =} \DecValTok{30}\NormalTok{)}
\end{Highlighting}
\end{Shaded}

\includegraphics{stats-gene-research-progress-v6_files/figure-latex/unnamed-chunk-3-3.pdf}

\end{document}
